File: metactical/metactical/report/item_details_report___v1/item_details_report___v1.js
// Copyright (c) 2024, Techlift Technologies and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Item Details Report - V1"] = {
	"filters": [
		{
			"fieldname":"supplier",
			"label": __("Supplier"),
			"fieldtype": "MultiSelectList",
			"options": "Supplier",
			"reqd": 1,
			get_data: function(txt) {
				return frappe.db.get_link_options("Supplier", txt);
			},
		},
		{
			"fieldname": "item_code",
			"label": __("Item Code"),
			"fieldtype": "Link",
			"options": "Item",
			get_query: function() {
				return {
					filters: {
						"has_variants": 0
					}
				};
			}
		},
		// number of records
		{
			"fieldname":"limit",
			"label": __("Limit"),
			"fieldtype": "Select",
			"options": ["20", "500", "1000", "5000", "10000", "All"],
			"default": "20"
		}
	]
};



// if filters.get("from_date"):
// conditions.append("po.transaction_date >= '{from_date}'".format(from_date=filters.get("from_date")))
// if filters.get("to_date"):
// conditions.append("po.transaction_date <= '{to_date}'".format(to_date=filters.get("to_date")))
// if filters.get("supplier"):
// conditions.append("supplier = '{supplier}'".format(supplier=filters.get("supplier")))
// if filters.get("item_code"):
// conditions.append("poi.item_code = '{item_code}'".format(item_code=filters.get("item_code")))
// if filters.get("retail_sku"):
// conditions.append("`tabItem`.ifw_retailskusuffix = '{retail_sku}'".format(retail_sku=filters.get("retail_sku")))

File: metactical/metactical/report/item_details_report___v1/item_details_report___v1.py
# Copyright (c) 2024, Techlift Technologies and contributors
# For license information, please see license.txt

import frappe

def execute(filters=None):
	columns = get_columns()
	conditions = get_conditions(filters)
	data = get_data(conditions, filters)

	return columns, data

def get_conditions(filters):
	conditions = ""
	suppliers = []
	limit = filters.get("limit")
	if filters.get('supplier'):
		suppliers = frappe.parse_json(filters.get("supplier"))
		conditions += " and supplier IN %(supplier)s"
	elif filters.get('item_code'):
		conditions += " and item_code = %(item_code)s"
	if limit != "All":
		conditions += " limit {}".format(str(limit))

	return conditions

def get_data(conditions, filters):
	data = frappe.db.sql("""
		SELECT
			item_code, description, variant_of, brand, supplier, supplier_part_no, default_price_list
		FROM
			`tabItem`
		LEFT JOIN
			`tabItem Supplier` ON `tabItem`.name = `tabItem Supplier`.parent
		LEFT JOIN
			`tabSupplier` ON `tabItem Supplier`.supplier = `tabSupplier`.name
		WHERE 1 = 1 %s
	"""%(conditions), filters, as_dict=1)

	for d in data:
		d["default_supplier_cost_price"] = frappe.db.get_value("Item Price", {"item_code": d["item_code"], "price_list": d["default_price_list"]}, "price_list_rate")
		d["standard_selling_price"] = frappe.db.get_value("Item Price", {"item_code": d["item_code"], "price_list": "RET - Camo"}, "price_list_rate")

	return data

def get_columns():
	# Item_code	Item_description	Item_Template	Item_Brand	Supplier_Code	Supplier_Name	Default Supplier Cost Price	Standard Selling Price
	return [
		{
			"label": "Item Code",
			"fieldname": "item_code",
			"fieldtype": "Link",
			"options": "Item",
			"width": 150
		},
		{
			"label": "Item Description",
			"fieldname": "description",
			"fieldtype": "Data",
			"width": 150
		},
		{
			"label": "Item Template",
			"fieldname": "variant_of",
			"fieldtype": "Data",
			"width": 150
		},
		{
			"label": "Item Brand",
			"fieldname": "item_brand",
			"fieldtype": "Data",
			"width": 150
		},
		{
			"label": "Supplier Part No",
			"fieldname": "supplier_part_no",
			"fieldtype": "Data",
			"width": 150
		},
		{
			"label": "Supplier Name",
			"fieldname": "supplier",
			"fieldtype": "Data",
			"width": 150
		},
		{
			"label": "Default Supplier Cost Price",
			"fieldname": "default_supplier_cost_price",
			"fieldtype": "Currency",
			"width": 150
		},
		{
			"label": "Standard Selling Price",
			"fieldname": "standard_selling_price",
			"fieldtype": "Currency",
			"width": 150
		}
	]
